src/pad/src/ICeWall.cpp

```diff
@@ -0,0 +1,1070 @@
+    return;
+  }
+
+  if (row == nullptr) {
+    return;
+  }
+  const double dbus = block->getDbUnitsPerMicron();
+
+  std::vector<odb::dbMaster*> fillers = masters;
+  // remove nullptrs
+  fillers.erase(std::remove(fillers.begin(), fillers.end(), nullptr),
+                fillers.end());
+  // sort by width
+  std::stable_sort(fillers.begin(),
+                   fillers.end(),
+                   [](odb::dbMaster* r, odb::dbMaster* l) -> bool {
+                     // sort biggest to smallest
+                     return r->getWidth() > l->getWidth();
+                   });
+
+  const odb::Rect rowbbox = row->getBBox();
+
+  using namespace boost::polygon::operators;
+  using Rectangle = boost::polygon::rectangle_data<int>;
+  using Polygon90 = boost::polygon::polygon_90_with_holes_data<int>;
+  using Polygon90Set = boost::polygon::polygon_90_set_data<int>;
+  using Pt = Polygon90::point_type;
+
+  std::vector<Polygon90> placed_io;
+  for (auto* inst : getPadInstsInRow(row)) {
+    const odb::Rect bbox = inst->getBBox()->getBox();
+    std::array<Pt, 4> pts;
+    if (row->getDirection() == odb::dbRowDir::HORIZONTAL) {
+      pts = {Pt(bbox.xMin(), rowbbox.yMin()),
+             Pt(bbox.xMax(), rowbbox.yMin()),
+             Pt(bbox.xMax(), rowbbox.yMax()),
+             Pt(bbox.xMin(), rowbbox.yMax())};
+    } else {
+      pts = {Pt(rowbbox.xMin(), bbox.yMin()),
+             Pt(rowbbox.xMax(), bbox.yMin()),
+             Pt(rowbbox.xMax(), bbox.yMax()),
+             Pt(rowbbox.xMin(), bbox.yMax())};
+    }
+
+    debugPrint(
+        logger_,
+        utl::PAD,
+        "Fill",
+        2,
+        "Instance in {} -> {} ({:.3f}um, {:.3f}um) -> ({:.3f}um, {:.3f}um)",
+        row->getName(),
+        inst->getName(),
+        pts[0].x() / dbus,
+        pts[0].y() / dbus,
+        pts[2].x() / dbus,
+        pts[2].x() / dbus);
+
+    Polygon90 poly;
+    poly.set(pts.begin(), pts.end());
+    placed_io.push_back(poly);
+  }
+
+  std::array<Pt, 4> pts = {Pt(rowbbox.xMin(), rowbbox.yMin()),
+                           Pt(rowbbox.xMax(), rowbbox.yMin()),
+                           Pt(rowbbox.xMax(), rowbbox.yMax()),
+                           Pt(rowbbox.xMin(), rowbbox.yMax())};
+
+  Polygon90 poly;
+  poly.set(pts.begin(), pts.end());
+  std::array<Polygon90, 1> arr{poly};
+
+  Polygon90Set new_shape(boost::polygon::HORIZONTAL, arr.begin(), arr.end());
+
+  for (const auto& io : placed_io) {
+    new_shape -= io;
+  }
+
+  std::vector<Rectangle> rects;
+  new_shape.get_rectangles(rects);
+
+  const int site_width = row->getSite()->getWidth();
+  int fill_group = 0;
+  for (auto& r : rects) {
+    const odb::Rect new_rect(xl(r), yl(r), xh(r), yh(r));
+
+    int width;
+    int start;
+    if (row->getDirection() == odb::dbRowDir::HORIZONTAL) {
+      width = new_rect.dx();
+      start = new_rect.xMin();
+    } else {
+      width = new_rect.dy();
+      start = new_rect.yMin();
+    }
+    int sites = width / site_width;
```

Comment:
What happens if the width isn't an integral site_width?  The round down could leave a gap.  If the assumption is that is an integral multiple then we should have a check.

Reply:
I added a check, but this ideally shouldn't happen since the pads a snapped to the site indexes.